jwst/ramp_fitting/ramp_fit.py

```diff
@@ -105,9 +118,286 @@ def ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
 
     return new_model, int_model, opt_model, gls_opt_model
 
+def ols_ramp_fit_multi(input_model, buffsize, save_opt, readnoise_2d, gain_2d,
+                 weighting, max_cores):
+    """
+       Setup the inputs to ols_ramp_fit with and without multiprocessing. The input inputs will be sliced into the
+       number of cores that are being used for multiprocessing. Because the data models cannot be pickeled, only
+       numpy arrays are passed and returned as parameters to ols_ramp_fit.
+
+       Parameters
+       ----------
+       input_model : data model
+           input data model, assumed to be of type RampModel
+
+       buffsize : int
+           size of data section (buffer) in bytes (not used)
+
+       save_opt : boolean
+          calculate optional fitting results
+
+       readnoise_model : instance of data Model
+           readnoise for all pixels
+
+       gain_model : instance of gain model
+           gain for all pixels
+
+       algorithm : string
+           'OLS' specifies that ordinary least squares should be used;
+           'GLS' specifies that generalized least squares should be used.
+
+       weighting : string
+           'optimal' specifies that optimal weighting should be used;
+            currently the only weighting supported.
+
+       max_cores : string
+           Number of cores to use for multiprocessing. If set to 'none' (the default), then no multiprocessing will be done.
+           The other allowable values are 'quarter', 'half', and 'all'. This is the fraction of cores to use for multi-proc.
+           The total number of cores includes the SMT cores (Hyper Threading for Intel).
+
+       Returns
+       -------
+       new_model : Data Model object
+           DM object containing a rate image averaged over all integrations in
+           the exposure
+
+       int_model : Data Model object or None
+           DM object containing rate images for each integration in the exposure
+
+       opt_model : RampFitOutputModel object or None
+           DM object containing optional OLS-specific ramp fitting data for the
+           exposure
+
+       gls_opt_model : GLS_RampFitModel object or None
+           Object containing optional GLS-specific ramp fitting data for the
+           exposure
+       """
+    if max_cores == 'none':
+        number_slices = 1
+    else:
+        num_cores = multiprocessing.cpu_count()
+        log.debug("Found %d possible cores to use for ramp fitting " % num_cores)
```

Comment:
future proof using `log.debug(f'Found {num_cores} possible cores to use for ramp fitting')`

Reply:
changed